ERCS/erc-7744.md

```diff
@@ -0,0 +1,101 @@
+---
+eip: 7744
+title: Code Index
+description: Indexing smart contract bytecode on Ethereum.
+author: Tim Pechersky (@peersky) <t@peersky.xyz>
+discussions-to: https://ethereum-magicians.org/t/erc-7744-code-index/20569
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-07-16
+---
+
+## Abstract
+
+This EIP defines a standard interface for indexing smart contracts on Ethereum by their bytecode hash. This enables trustless discovery and verification of contract code, facilitating use cases like bytecode signing, whitelisting, and decentralized distribution mechanisms.
+
+## Motivation
+
+Existing contract discovery relies on addresses, which are non-deterministic and can be obfuscated through proxies. Indexing by bytecode hash provides a deterministic and tamper-proof way to identify and verify contract code, enhancing security and trust in the Ethereum ecosystem.
+
+Global bytecode identification is a requirement to build efficient at high scale global factories for future software distribution, that may want to rely on extensive re-use of same codebase as own dependencies.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+```solidity
+// SPDX-License-Identifier: CC0-1.0
+pragma solidity =0.8.20;
+
+interface ICodeIndex {
+    event Indexed(address indexed container, bytes32 indexed codeHash);
+    error alreadyExists(bytes32 id, address source);
+
+    function register(address container) external;
+
+    function get(bytes32 id) external view returns (address);
+}
+
+/**
+ * @title Byte Code Indexer Contract
+ * @notice You can use this contract to index contracts by their bytecode.
+ * @dev This allows to query contracts by their bytecode instead of addresses.
+ * @author Tim Pechersky (@Peersky)
+ */
+contract CodeIndex is ICodeIndex {
+    mapping(bytes32 => address) private index;
+
+    /**
+     * @notice Registers a contract in the index by its bytecode hash
+     * @param container The contract to register
+     * @dev `msg.codeHash` will be used
+     * @dev It will revert if the contract is already indexed
+     */
+    function register(address container) external {
+        address etalon = index[container.codehash];
+        if (etalon != address(0)) {
+            revert alreadyExists(container.codehash, etalon);
+        }
+        index[container.codehash] = container;
+        emit Indexed(container, container.codehash);
+    }
+
+    /**
+     * @notice Returns the contract address by its bytecode hash
+     * @dev returns zero if the contract is not indexed
+     * @param id The bytecode hash
+     * @return The contract address
+     */
+    function get(bytes32 id) external view returns (address) {
+        return index[id];
+    }
+}
+
+```
```

Comment:
We generally don't allow implementations in the specification section. We have made exceptions for registries in the past, but you'll need to explain (in the document) why this registry needs to be an EIP at all, and not just a library you publish.

Reply:
added motivation